cloud/libvirt/actuators/machine/utils/domain.go

```diff
@@ -745,3 +754,125 @@ func DomainExists(machine *clusterv1.Machine) (bool, error) {
 
 	return true, nil
 }
+
+// LookupDomainByUUID retrieves a libvirt domain by UUID using the
+// provider config on the given machine.
+func LookupDomainByUUID(machine *clusterv1.Machine, uuid *uuid.UUID) (*libvirt.Domain, error) {
+	machineProviderConfig, err := MachineProviderConfigFromClusterAPIMachineSpec(&machine.Spec)
+	client, err := buildClient(machineProviderConfig.URI)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to build libvirt client: %s", err)
+	}
+
+	virConn := client.libvirt
+	if virConn == nil {
+		return nil, fmt.Errorf(LibVirtConIsNil)
+	}
+
+	dom, err := virConn.LookupDomainByUUID(uuid[:])
+	if err != nil {
+		return nil, fmt.Errorf("Failed to lookup domain by UUID: %v", err)
+	}
+
+	return dom, nil
+}
+
+// ProviderStatus deserializes a libvirt provider status from a
+// machine object.
+func ProviderStatus(machine *clusterv1.Machine) (*providerconfigv1.LibvirtMachineProviderStatus, error) {
+	if machine.Status.ProviderStatus == nil {
+		return &providerconfigv1.LibvirtMachineProviderStatus{}, nil
+	}
+
+	decoder := providerconfigv1.Codecs.UniversalDecoder(providerconfigv1.SchemeGroupVersion)
+	obj, gvk, err := decoder.Decode([]byte(machine.Status.ProviderStatus.Raw), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	status, ok := obj.(*providerconfigv1.LibvirtMachineProviderStatus)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object: %#v", gvk)
```

Comment:
Would expressing `%T` be more useful here? i.e., what type did we get?

Reply:
This `gvk` struct will have the group, version, and kind information. Honestly, this bit is just lifted from the AWS actuator, but it seems like a reasonable thing to log in this situation.